From d060ac32b3d2f3fa7925bcc099cdf2d570c68939 Mon Sep 17 00:00:00 2001
From: zachary <zhangzg@marvell.com>
Date: Mon, 13 Mar 2017 23:02:41 +0800
Subject: [PATCH 0922/1345] phy: comphy: a3700: add USB3 phy power off support

commit  9c33284e9f2703ab5b4ad1a2d7cb88af3f2ffe42 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

- The USB3 MAC will control the USB3 PHY to put it into low power
  state when necessary. Thus do not need to power off the USB3 PHY
  again.
- However need to register the USB3 PHY power off routine to avoid
  the error while the USB3 driver tries to power off the PHY, which
  is a generic way for power saving.

Change-Id: I708f7eb5569b1ce72151eb3266c67567b8a94615
Signed-off-by: zachary <zhangzg@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/37393
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Victor Gu <xigu@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/phy/phy-comphy-a3700.c |   17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/drivers/phy/phy-comphy-a3700.c b/drivers/phy/phy-comphy-a3700.c
index d631f14..00a591a 100644
--- a/drivers/phy/phy-comphy-a3700.c
+++ b/drivers/phy/phy-comphy-a3700.c
@@ -728,6 +728,20 @@ static int mvebu_a3700_comphy_usb3_power_on(struct mvebu_comphy_priv *priv,
 	return ret;
 }
 
+static int mvebu_a3700_comphy_usb3_power_off(struct mvebu_comphy_priv *priv,
+					    struct mvebu_comphy *comphy)
+{
+	/*
+	 * Currently the USB3 MAC will control the USB3 PHY to set it to low state,
+	 * thus do not need to power off USB3 PHY again.
+	 */
+	dev_dbg(priv->dev, "%s: Enter\n", __func__);
+
+	dev_dbg(priv->dev, "%s: Exit\n", __func__);
+
+	return 0;
+}
+
 static int mvebu_a3700_comphy_pcie_power_on(struct mvebu_comphy_priv *priv,
 					    struct mvebu_comphy *comphy)
 {
@@ -878,6 +892,9 @@ static int mvebu_a3700_comphy_power_off(struct phy *phy)
 	spin_lock(&priv->lock);
 
 	switch (mode) {
+	case (COMPHY_USB3_MODE):
+		err = mvebu_a3700_comphy_usb3_power_off(priv, comphy);
+		break;
 	case (COMPHY_SATA_MODE):
 		err = mvebu_a3700_comphy_sata_power_off(priv, comphy);
 		break;
-- 
1.7.9.5

